### PR TITLE
INTMDB-939: [Terraform] Feature Add: Shared-Tier Snapshots

### PR DESCRIPTION
**Pull request by andreaangiolillo**
## Description

Ticket: [INTMDB-939](https://jira.mongodb.org/browse/INTMDB-939)
This PR adds support to `mongodbatlas_shared_tier_snapshot` and `mongodbatlas_shared_tier_snapshots`

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [x] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contribution guidelines](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/CONTRIBUTING.md)
- [x] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run make fmt and formatted my code

### Tested Locally
```terraform
terraform state show data.mongodbatlas_shared_tier_snapshot.test

# data.mongodbatlas_shared_tier_snapshot.test:
data "mongodbatlas_shared_tier_snapshot" "test" {
    cluster_name     = "Cluster0"
    expiration       = "2023-07-26 07:30:47 +0000 UTC"
    finish_time      = "2023-07-18 07:33:54 +0000 UTC"
    id               = "64b63fb50288d52922020d24"
    mongo_db_version = "6.0.8"
    project_id       = "63dcfc256af00a5934e60924"
    scheduled_time   = "2023-07-18 07:30:47 +0000 UTC"
    snapshot_id      = "64b63fb50288d52922020d24"
    start_time       = "2023-07-18 07:31:07 +0000 UTC"
    status           = "COMPLETED"
}
```

```terraform
terraform state show data.mongodbatlas_shared_tier_snapshots.test
# data.mongodbatlas_shared_tier_snapshots.test:
data "mongodbatlas_shared_tier_snapshots" "test" {
    cluster_name = "Cluster0"
    id           = "terraform-20230718091336580300000001"
    project_id   = "63dcfc256af00a5934e60924"
    results      = [
        {
            expiration       = "2023-07-26 07:30:47 +0000 UTC"
            finish_time      = "2023-07-18 07:33:54 +0000 UTC"
            mongo_db_version = "6.0.8"
            scheduled_time   = "2023-07-18 07:30:47 +0000 UTC"
            snapshot_id      = "64b63fb50288d52922020d24"
            start_time       = "2023-07-18 07:31:07 +0000 UTC"
            status           = "COMPLETED"
        },
    ]
    total_count  = 1
}
```

### CODE REVIEW

File: website/docs/d/cloud_provider_shared_tier_snapshot.html.markdown [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+description: |-
+    Provides an Cloud Backup Snapshot Datasource for Shared Tier Clusters.
+---
+
+# Data Source: mongodbatlas_shared_tier_snapshot
+
+`mongodbatlas_shared_tier_snapshot` provides an Cloud Backup Snapshot datasource for Shared Tier Clusters. Atlas Cloud Backup Snapshots provide localized backup storage using the native snapshot functionality of the cluster’s cloud service.
+
+-> **NOTE:** Groups and projects are synonymous terms. You may find `groupId` in the official documentation.
+-> **NOTE:** This datasource is only for Shared Tier Clusters (M0, M2, M5).
```
Comment (Zuhairahmed) [inline website/docs/d/cloud_provider_shared_tier_snapshot.html.markdown]:
```suggestion
-> **NOTE:** This data source is only for Shared Tier Clusters (M2 and M5). See [here](https://www.mongodb.com/docs/atlas/reference/free-shared-limitations/) for more details on Shared Tier Cluster Limitations. 
```

File: website/docs/d/cloud_provider_shared_tier_snapshot.html.markdown [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+layout: "mongodbatlas"
+page_title: "MongoDB Atlas: mongodbatlas_shared_tier_snapshot"
+sidebar_current: "docs-mongodbatlas-datasource-cloud_provider_shared_tier_snapshot"
+description: |-
+    Provides an Cloud Backup Snapshot Datasource for Shared Tier Clusters.
+---
+
+# Data Source: mongodbatlas_shared_tier_snapshot
+
+`mongodbatlas_shared_tier_snapshot` provides an Cloud Backup Snapshot datasource for Shared Tier Clusters. Atlas Cloud Backup Snapshots provide localized backup storage using the native snapshot functionality of the cluster’s cloud service.
```
Comment (Zuhairahmed) [inline website/docs/d/cloud_provider_shared_tier_snapshot.html.markdown]:
```suggestion
`mongodbatlas_shared_tier_snapshot` provides an Cloud Backup Snapshot data source for Shared Tier Clusters. Atlas Cloud Backup Snapshots provide localized backup storage using the native snapshot functionality of the cluster’s cloud service.
```

File: website/docs/d/cloud_provider_shared_tier_snapshots.html.markdown [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+description: |-
+    Provides a Cloud Backup Snapshots Datasource for Shared Tier Clusters.
+---
+
+# Data Source: mongodbatlas_shared_tier_snapshots
+
+`mongodbatlas_shared_tier_snapshots` provides an Cloud Backup Snapshots datasource for Shared Tier Clusters. Atlas Cloud Backup Snapshots provide localized backup storage using the native snapshot functionality of the cluster’s cloud service.
+
+-> **NOTE:** Groups and projects are synonymous terms. You may find `groupId` in the official documentation.
+-> **NOTE:** This datasource is only for Shared Tier Clusters (M0, M2, M5).
```
Comment (Zuhairahmed) [inline website/docs/d/cloud_provider_shared_tier_snapshots.html.markdown]:
```suggestion
-> **NOTE:** This data source is only for Shared Tier Clusters (M2 and M5). See [here](https://www.mongodb.com/docs/atlas/reference/free-shared-limitations/) for more details on Shared Tier Cluster Limitations. 
```

File: website/docs/d/cloud_provider_shared_tier_snapshots.html.markdown [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+layout: "mongodbatlas"
+page_title: "MongoDB Atlas: mongodbatlas_shared_tier_snapshots"
+sidebar_current: "docs-mongodbatlas-datasource-cloud_provider_shared_tier_snapshots"
+description: |-
+    Provides a Cloud Backup Snapshots Datasource for Shared Tier Clusters.
+---
+
+# Data Source: mongodbatlas_shared_tier_snapshots
+
+`mongodbatlas_shared_tier_snapshots` provides an Cloud Backup Snapshots datasource for Shared Tier Clusters. Atlas Cloud Backup Snapshots provide localized backup storage using the native snapshot functionality of the cluster’s cloud service.
```
Comment (Zuhairahmed) [inline website/docs/d/cloud_provider_shared_tier_snapshots.html.markdown]:
```suggestion
`mongodbatlas_shared_tier_snapshots` provides an Cloud Backup Snapshots data source for Shared Tier Clusters. Atlas Cloud Backup Snapshots provide localized backup storage using the native snapshot functionality of the cluster’s cloud service.
```

**Review by Zuhairahmed — APPROVED** [review]
Few doc comments, otherwise LGTM

File: mongodbatlas/data_source_mongodbatlas_shared_tier_snapshot.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+// This datasource does not have a resource: we tested it manually
+func dataSourceMongoDBAtlasSharedTierSnapshot() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: dataSourceMongoDBAtlasSharedTierSnapshotRead,
+		Schema: map[string]*schema.Schema{
```
Comment (wtrocki) [inline mongodbatlas/data_source_mongodbatlas_shared_tier_snapshot.go:15]: FYI - We could easily generate initial schemas for developers to adapt them in the future. 
We have that capability in the SDK - just need to adjust it to your needs.

**Review by wtrocki — APPROVED** [review]

File: mongodbatlas/data_source_mongodbatlas_shared_tier_snapshot.go (hunk 5 above)
Comment (andreaangiolillo) [reply mongodbatlas/data_source_mongodbatlas_shared_tier_snapshot.go:15]: Could you share more details about it? A link to the SDK codebase where I can look at would be enough. Thanks!

**Review by gssbzn — APPROVED** [review]

File: mongodbatlas/data_source_mongodbatlas_shared_tier_snapshot.go (hunk 5 above)
Comment (wtrocki) [reply mongodbatlas/data_source_mongodbatlas_shared_tier_snapshot.go:15]: I missed that comment.
We have a doc for using custom templates:

https://github.com/mongodb/atlas-sdk-go/blob/main/tools/CONTRIBUTING.md#generating-custom-files

We can generate some level of boilerplate for the team like TF schemas.
We can release it under separate module or package.

**Pull request merged**